src/Utils/negative_tracers.jl

```diff
@@ -135,33 +141,74 @@ function update_biogeochemical_state!(model, scale::ScaleNegativeTracers)
 
     tracers_to_scale = Tuple(model.tracers[tracer_name] for tracer_name in scale.tracers)
 
-    scale_for_negs_kernel!(tracers_to_scale, scale.scalefactors, scale.invalid_fill_value)
+    field_scale = Tuple(Iterators.flatten(
+        zip([model.tracers[tracer_name] for tracer_name in scale.tracers],
+            scale.scalefactors
+        )))
+
+    scale_for_negs_kernel!(scale.invalid_fill_value, field_scale...)
 
     return nothing
 end
 
-@kernel function scale_for_negs!(tracers, scalefactors, invalid_fill_value)
-    i, j, k = @index(Global, NTuple)
-
-    t, p = 0.0, 0.0
-
-    for (idx, tracer) in enumerate(tracers)
-        value = @inbounds tracer[i, j, k]
-        scalefactor = @inbounds scalefactors[idx]
-
-        t += value * scalefactor
-        if value > 0
-            p += value * scalefactor
-        end
-    end
+# `CUDA.jl` has a tendency to make a local copy if we pass a list of tracer fields
+# as a tuple (despite it beeing immutable). See the related issue:
+# https://github.com/JuliaGPU/CUDA.jl/issues/1168
+#
+# This makes each thread use a large amout of 'Local' memory, which in turn 
+# introduces large amout of unoptimal memory traffic and slows the kernel
+# singnificantly.
+#
+# The easiest workaround to avoid the local copy is to pass each field (and
+# associated scalfactor) as a parameter to the kernel. But since we want to
+# support arbitrary number of fields we need to make the kernel variadic.
+#
+# We expect Julia to inline the recursive calls and, effectivly unroll the loops
+@kernel function scale_for_negs!(invalid_fill_value, field_scale...)
+    ijk = @index(Global, NTuple)
+
+    t, p = calculate_total_and_positive_part(0.0, 0.0, ijk, field_scale...)
 
     t = ifelse(t < 0, invalid_fill_value, t)
 
-    for tracer in tracers
-        value = @inbounds tracer[i, j, k]
+    correct_negative_fields!(t, p, ijk, field_scale...)
+    nothing
+end
 
-        new_value = ifelse(!isfinite(value) | (value > 0), value * t / p, 0)
+# Recursive step
+@inline function calculate_total_and_positive_part(t,
+                                                   p,
+                                                   ijk,
+                                                   field,
+                                                   scale,
+                                                   field_scale...)
+    t, p = calculate_total_and_positive_part(t, p, ijk, field, scale)
+    return calculate_total_and_positive_part(t, p, ijk, field_scale...)
+end
 
-        @inbounds tracer[i, j, k] = new_value
+# Recursion terminal
+@inline function calculate_total_and_positive_part(t, p, ijk, field, scale)
+    i, j, k = ijk
+    value = @inbounds field[i, j, k]
+    t += value * scale
+    if value > 0
+        p += value * scale
```

Comment:
I am lead to believe that it avoids branching if we were instead to write:
```suggestion
p = ifelse(value>0, p + value * scale, p)
```